### PR TITLE
feat: support control parallel build count from env

### PR DESCRIPTION
**Pull request by syu-w**
Support use the following environment variables to control the parrallel build count:

Use
`<appname>_PARALLEL_BUILD_COUNT`
`CRAFT_PARALLEL_BUILD_COUNT`
to set to a fixed value.

Use
`<appname>_MAX_PARALLEL_BUILD_COUNT`
`CRAFT_MAX_PARALLEL_BUILD_COUNT`
to set to a dynamic value based on cpu count and max value.

By default it will use the current CPU count value to the parrallel build count.

Fixes: https://github.com/canonical/craft-application/issues/180

### CODE REVIEW

File: craft_application/services/lifecycle.py [hunk 1]
```diff
@@ -265,3 +267,61 @@ def __repr__(self) -> str:
...
+        effect_env_name = "PARALLEL BUILD COUNT"
+
+        # fixed parallel build count environment variable
+        for env_name in [
+            (self._app.name + "_PARALLEL_BUILD_COUNT").upper(),
+            "CRAFT_PARALLEL_BUILD_COUNT",
+        ]:
+            if os.environ.get(env_name):
+                try:
+                    parallel_build_count = int(os.environ[env_name])
```
Comment (mr-cal) [inline craft_application/services/lifecycle.py:283]: Do you mind adding a debug statement to inform the user how the value was determined?

Like `emit.debug(f"Using parallel build count of {parallel_build_count} from environment variable {env_name!r}")`?

And probably two more debug statements later on where `parallel_build_count` is set.

File: craft_application/services/lifecycle.py [hunk 2]
```diff
@@ -265,3 +267,61 @@ def __repr__(self) -> str:
             f"{self.__class__.__name__}({self._app!r}, {self._project!r}, "
             f"{work_dir=}, {cache_dir=}, {build_for=}, **{self._manager_kwargs!r})"
         )
+
+    def _get_parallel_build_count(self) -> int:  # noqa: PLR0912
```
Comment (mr-cal) [inline craft_application/services/lifecycle.py:271]: Can you add a docstring describing how the parallel build count is determined?

**Review by mr-cal — APPROVED** [review]
Looks good, just added some requests for docstrings and logging.

The following opinion doesn't have to be done, since the code does what it is supposed to do and is self-contained, but I'll share it anyways:

I agree with ruff (seeing the disabled `PLR0912` rule on the function) that this could be written more clearly.  One possibility would be to move the ` if os.environ.get(env_name): ... try/except` into a private helper function.  The validation of the values being positive integers could probably be moved in there as well and may allow getting rid of another 1-2 if statements.

**Comment by syu-w** [comment]
> Looks good, just added some requests for docstrings and logging.
> 
> The following opinion doesn't have to be done, since the code does what it is supposed to do and is self-contained, but I'll share it anyways:
> 
> I agree with ruff (seeing the disabled `PLR0912` rule on the function) that this could be written more clearly. One possibility would be to move the ` if os.environ.get(env_name): ... try/except` into a private helper function. The validation of the values being positive integers could probably be moved in there as well and may allow getting rid of another 1-2 if statements.

Updated based on feedback

**Review by mr-cal — APPROVED** [review]
Looks great, thanks!

**Review by lengau — APPROVED** [review]
Looks good, thanks!

**Pull request merged**